orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishDeliveryConfigTask.kt

```diff
@@ -0,0 +1,157 @@
+import com.netflix.spinnaker.orca.RetryableTask
+import com.netflix.spinnaker.orca.TaskResult
+import com.netflix.spinnaker.orca.igor.ScmService
+import com.netflix.spinnaker.orca.pipeline.model.GitTrigger
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Component
+import retrofit.RetrofitError
+import java.util.concurrent.TimeUnit
+
+/**
+ * Task that retrieves a Managed Delivery config manifest from source control via igor, then publishes it to keel,
+ * to support GitOps flows.
+ */
+@Component
+class PublishDeliveryConfigTask
+constructor(
+  private val keelService: KeelService,
+  private val scmService: ScmService,
+  private val objectMapper: ObjectMapper
+) : RetryableTask {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  override fun execute(stage: Stage): TaskResult {
+    val context = objectMapper.convertValue<PublishDeliveryConfigContext>(stage.context)
+    val trigger = stage.execution.trigger
+    val user = trigger.user ?: "anonymous"
+
+    if (context.attempt > context.maxRetries!!) {
+      val error = "Maximum number of retries exceeded (${context.maxRetries})"
+      log.error("$error. Aborting.")
+      return TaskResult.builder(ExecutionStatus.TERMINAL).context(mapOf("error" to error)).build()
+    }
+
+    // if the pipeline has a git trigger, infer what context we can from the trigger
+    if (trigger is GitTrigger) {
+      // FIXME: GitTrigger props are non-null, but IIUC branch and hash would be mutually exclusive, so what to do here?
+      if (trigger.hash != null && context.ref == null) {
+        context.ref = trigger.hash
+        log.debug("Inferred context.ref from trigger: ${context.ref}")
+      } else if (trigger.branch != null && context.ref == null) {
+        context.ref = "refs/heads/${trigger.branch}"
+        log.debug("Inferred context.ref from trigger: ${context.ref}")
+      }
+      if (context.scmType == null) {
+        context.scmType = trigger.source
+        log.debug("Inferred context.scmType from trigger: ${context.scmType}")
+      }
+      if (context.project == null) {
+        context.project = trigger.project
+        log.debug("Inferred context.project from trigger: ${context.project}")
+      }
+      if (context.repository == null) {
+        context.repository = trigger.slug
+        log.debug("Inferred context.repository from trigger: ${context.repository}")
+      }
+    } else {
+      if (context.ref == null) {
+        context.ref = "refs/heads/master"
+      }
+      if (context.scmType == null || context.project == null || context.repository == null) {
+        throw IllegalArgumentException("scmType, project and repository are required fields in the stage if there's no git trigger.")
+      }
+    }
+
+    val manifestLocation = "${context.scmType}://${context.project}/${context.repository}/.netflix/${context.directory
```

Comment:
`.netflix` shouldn't be hardcoded here, it should be a property value that defaults to `.spinnaker` that we can override internally

Reply:
@emjburns I was thinking about that and I think the right way to do this would be for `igor` to expose this config via an API, instead of duplicating it here, since it's truly only used for logging here.